File: api/services/runner.py
"""Run the C++ simulation pipeline in a worker thread."""

from __future__ import annotations

import json
import traceback
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from services import cpp_adapter
from services.storage import outputs_root

PROGRESS: dict[str, dict[str, Any]] = {}


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def set_progress(build: str, phase: str, pct: float, message: str, *, done: bool = False) -> None:
    PROGRESS[build] = {
        "phase": phase,
        "pct": max(0.0, min(100.0, float(pct))),
        "message": message,
        "done": done,
    }


def clear_progress_later(build: str) -> None:
    """Caller may remove PROGRESS after SSE clients disconnect; optional."""
    PROGRESS.pop(build, None)


def _merge_metadata(build: str, updates: dict[str, Any]) -> None:
    out = outputs_root()
    path = out / build / "metadata.json"
    existing: dict[str, Any] = {}
    if path.is_file():
        try:
            existing = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            existing = {}
    existing.update(updates)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(existing, indent=2), encoding="utf-8")


def run_simulation_job(
    build: str,
    export_path: Path,
    teaminfo_path: Path,
    teams: list[str],
    seed: int,
    runs: int,
    n_workers: int,
) -> None:
    """Execute the full pipeline via the C++ engine adapter.

    Assumes metadata.json already exists with status running.
    """
    canonical_run_dir = outputs_root() / build
    canonical_run_dir.mkdir(parents=True, exist_ok=True)

    try:
        set_progress(build, "parsing", 2.0, "Loading export…")
        _merge_metadata(build, {"status": "running"})

        # Phase weights: parsing 0–5, simulating 5–80, saving/analyzing 80–97,
        # finalizing 97–99, complete 100.
        set_progress(build, "simulating", 5.0, f"Simulating ({runs} runs)…")

        def _on_cpp_progress(cpp_pct: float, message: str) -> None:
            ui_pct = 5.0 + (cpp_pct * 0.75)
            set_progress(build, "simulating", ui_pct, f"Simulating ({int(cpp_pct)}%)")

        _stage_map = {
            "cpp_done": (80.0, "analyzing", "Saving workbook…"),
            "artifacts_copied": (85.0, "analyzing", "Copied artifacts."),
            "analyzing": (88.0, "analyzing", "Generating analysis…"),
            "analysis_done": (97.0, "analyzing", "Analysis complete."),
        }

        def _on_post_sim_stage(stage: str, message: str) -> None:
            mapping = _stage_map.get(stage)
            if mapping is None:
                return
            pct, phase, default_msg = mapping
            set_progress(build, phase, pct, message or default_msg)

        player_count = cpp_adapter.run_cpp_simulation(
            export_path=export_path,
            teaminfo_path=teaminfo_path,
            teams=teams,
            seed=seed,
            runs=runs,
            n_workers=n_workers,
            canonical_run_dir=canonical_run_dir,
            progress_callback=_on_cpp_progress,
            stage_callback=_on_post_sim_stage,
        )

        set_progress(build, "finalizing", 99.0, "Writing metadata…")
        _merge_metadata(
            build,
            {
                "status": "complete",
                "completed_at": _utc_now_iso(),
                "player_count": player_count,
                "error": None,
            },
        )
        set_progress(build, "complete", 100.0, "Done.", done=True)
    except Exception as exc:  # noqa: BLE001 — surface engine failures to metadata
        err = f"{type(exc).__name__}: {exc}"
        tb = traceback.format_exc()
        _merge_metadata(
            build,
            {
                "status": "failed",
                "completed_at": _utc_now_iso(),
                "error": err,
            },
        )
        set_progress(build, "failed", 0.0, err, done=True)
        _merge_metadata(build, {"error_detail": tb[:8000]})

File: api/src/services/runner.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import { outputsRoot } from "../paths.js";
import * as cppAdapter from "./cppAdapter.js";
import { setProgress } from "./progress.js";

export { PROGRESS, setProgress } from "./progress.js";

function utcNowIso(): string {
  return new Date().toISOString();
}

async function mergeMetadata(build: string, updates: Record<string, unknown>): Promise<void> {
  const out = outputsRoot();
  const metaPath = path.join(out, build, "metadata.json");
  let existing: Record<string, unknown> = {};
  try {
    if (fs.existsSync(metaPath)) {
      existing = JSON.parse(await fsp.readFile(metaPath, "utf8")) as Record<string, unknown>;
    }
  } catch {
    existing = {};
  }
  Object.assign(existing, updates);
  await fsp.mkdir(path.dirname(metaPath), { recursive: true });
  await fsp.writeFile(metaPath, JSON.stringify(existing, null, 2), "utf8");
}

/**
 * Patch canonical metadata from the engine's own `engine_metadata.json` — the
 * source of truth for what actually ran (progression id/name + sim params).
 * No-op if the engine file is missing/unreadable.
 */
async function patchFromEngineMetadata(build: string, runDir: string): Promise<void> {
  const enginePath = path.join(runDir, "engine_metadata.json");
  let engine: Record<string, unknown>;
  try {
    engine = JSON.parse(await fsp.readFile(enginePath, "utf8")) as Record<string, unknown>;
  } catch {
    return;
  }
  const prog = (engine.progression ?? {}) as Record<string, unknown>;
  const sim = (engine.simulation ?? {}) as Record<string, unknown>;
  const updates: Record<string, unknown> = {};
  const progId = typeof prog.id === "string" ? prog.id : undefined;
  const progName = typeof prog.name === "string" ? prog.name : undefined;
  if (progId || progName) updates.progression = { id: progId ?? null, name: progName ?? null };
  if (progName || progId) updates.script_version = progName ?? progId;
  if (typeof sim.seed === "number") updates.seed = sim.seed;
  if (typeof sim.runs === "number") updates.runs = sim.runs;
  if (typeof sim.workers === "number") updates.n_workers = sim.workers;
  if (typeof sim.year === "number") updates.year = sim.year;
  if (Object.keys(updates).length > 0) await mergeMetadata(build, updates);
}

export async function runSimulationJob(
  build: string,
  exportPath: string,
  teaminfoPath: string,
  teams: string[],
  seed: number,
  runs: number,
  n_workers: number,
  version = "v43",
): Promise<void> {
  const canonicalRunDir = path.join(outputsRoot(), build);
  await fsp.mkdir(canonicalRunDir, { recursive: true });

  try {
    setProgress(build, "parsing", 2, "Loading export…");
    await mergeMetadata(build, { status: "running" });

    setProgress(build, "simulating", 5, `Simulating (${runs} runs)…`);

    const onCppProgress = (cppPct: number, _message: string) => {
      const uiPct = 5 + cppPct * 0.75;
      setProgress(build, "simulating", uiPct, `Simulating (${Math.floor(cppPct)}%)`);
    };

    const stageMap: Record<string, [number, string, string]> = {
      cpp_done: [80, "analyzing", "Saving workbook…"],
      artifacts_copied: [85, "analyzing", "Copied artifacts."],
      analyzing: [88, "analyzing", "Generating analysis…"],
      analysis_done: [97, "analyzing", "Analysis complete."],
    };

    const onPostSimStage = (stage: string, message: string) => {
      const mapping = stageMap[stage];
      if (!mapping) return;
      const [pct, phase, defaultMsg] = mapping;
      setProgress(build, phase, pct, message || defaultMsg);
    };

    const { playerCount, analysisEngine } = await cppAdapter.runCppSimulation({
      exportPath,
      teaminfoPath,
      teams,
      seed,
      runs,
      n_workers,
      canonicalRunDir,
      version,
      progressCallback: onCppProgress,
      stageCallback: onPostSimStage,
    });

    setProgress(build, "finalizing", 99, "Writing metadata…");
    await patchFromEngineMetadata(build, canonicalRunDir);
    await mergeMetadata(build, {
      status: "complete",
      completed_at: utcNowIso(),
      player_count: playerCount,
      analysis_engine: analysisEngine,
      error: null,
    });
    setProgress(build, "complete", 100, "Done.", true);
  } catch (exc: unknown) {
    const err = exc instanceof Error ? exc : new Error(String(exc));
    const tb = err.stack ?? String(exc);
    await mergeMetadata(build, {
      status: "failed",
      completed_at: utcNowIso(),
      error: `${err.name}: ${err.message}`,
    });
    setProgress(build, "failed", 0, `${err.name}: ${err.message}`, true);
    await mergeMetadata(build, { error_detail: tb.slice(0, 8000) });
  }
}
